test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java

```diff
@@ -0,0 +1,339 @@
+		String myClasspath = System.getProperty("java.class.path");
+		argBuffer.add(javaExec);
+		argBuffer.add("-Dcom.ibm.tools.attach.enable=yes");
+		if (doLogging) {
+			argBuffer.add("-Dcom.ibm.tools.attach.logging=yes");
+		}
+		if (null != tgtId) {
+			argBuffer.add("-Dcom.ibm.tools.attach.id=" + tgtId);
+		}
+		if (null != myDisplayName) {
+			argBuffer.add("-Dcom.ibm.tools.attach.displayName=" + myDisplayName);
+		}
+		argBuffer.add("-classpath");
+		argBuffer.add(myClasspath);
+		if ((null != sideCar) && (sideCar.length() > 0)) {
+			String sidecarArgs[] = sideCar.split(" +");
+			for (String s : sidecarArgs) {
+				argBuffer.add(s);
+			}
+		}
+		if (null != vmArgs) {
+			argBuffer.addAll(vmArgs);
+		}
+		argBuffer.add(cmdName);
+		if (null != appArgs) {
+			argBuffer.addAll(appArgs);
+		}
+		Process target = null;
+		args = new String[argBuffer.size()];
+		argBuffer.toArray(args);
+		if (verbose) {
+			System.out.print("\n");
+			for (int i = 0; i < args.length; ++i) {
+				System.out.print(args[i] + " ");
+			}
+			System.out.print("\n");
+		}
+		try {
+			target = me.exec(args);
+		} catch (IOException e) {
+			e.printStackTrace();
+			return null;
+		}
+		this.targetIn = target.getOutputStream();
+		InputStream targetOut = target.getInputStream();
+		InputStream targetErr = target.getErrorStream();
+		targetInWriter = new BufferedWriter(new OutputStreamWriter(targetIn));
+		targetOutReader = new BufferedReader(new InputStreamReader(targetOut));
+		targetErrReader = new BufferedReader(new InputStreamReader(targetErr));
+		try {
+			targetInWriter.write(TargetManager.TARGETVM_START + '\n');
+			targetInWriter.flush();
+		} catch (IOException e) {
+			e.printStackTrace();
+			return null;
+		}
+		return target;
+	}
+
+	static void listIpcDir() {
+		String tmpdir = System.getProperty("java.io.tmpdir")
+				+ File.separatorChar + DEFAULT_IPC_DIR;
+		logger.debug("DEBUG:" + tmpdir);
+		File ipcDir = new File(tmpdir);
+		if (ipcDir.exists()) {
+			listRecursive(System.getProperty("java.io.tmpdir"), ipcDir);
+		}
+	}
+
+	static void listRecursive(String prefix, File root) {
+		String myPath = prefix + "/" + root.getName();
+		logger.debug(myPath);
+		if (root.isDirectory()) {
+			File[] children = root.listFiles();
+			if (null != children) {
+				for (File c : children) {
+					listRecursive(myPath, c);
+				}
+			}
+		}
+	}
+
+	protected int terminateTarget() {
+		int rc = -1;
+		if (!active) {
+			return 0;
+		}
+		active = false;
+		try {
+			targetInWriter.write(TargetManager.TARGETVM_STOP + '\n');
+			try {
+				targetInWriter.flush();
+			} catch (IOException e) {
+				/* ignore it */
+			}
+			while (targetErrReader.ready()) {
+				String currentLine = targetErrReader.readLine();
+				if ((null != currentLine) &&
+						!currentLine.startsWith("JVMJ9VM082E Unable to switch to IFA processor")) {
+					errOutput += currentLine + "\n";
+				}
+			}
+			while (targetOutReader.ready()) {
```

Comment:
I don't think this is the right code.   I would expect a loop while readLine() is not returning null.

Reply:
As mentioned above, I don't want readline() to block.

Reply:
I don't think this actually stops readline from blocking - the stream is ready as soon as there are characters available.

Reply:
The readline() will block until there is a newline, but the target process emits complete lines, so any blocking will be brief.  The purpose of the code was to handle the case where stdout would close and stderr was still open but not receiving anything.